cmd/root_test.go

```diff
@@ -64,18 +66,70 @@ func TestMain(m *testing.M) {
 	exitCode = m.Run()
 }
 
+type bufferStringer interface {
+	io.ReadWriter
+	fmt.Stringer
+	Bytes() []byte
+}
+
 type globalTestState struct {
 	*globalState
 	cancel func()
 
-	stdOut, stdErr *bytes.Buffer
+	stdOut, stdErr bufferStringer
 	loggerHook     *testutils.SimpleLogrusHook
 
 	cwd string
 
 	expectedExitCode int
 }
 
+// A thread-safe buffer implementation.
+type safeBuffer struct {
```

Comment:
As I wrote in the other comment, I think this is wrong. We don't want a safe and synchronized buffer here, we want the exact opposite. We want tests to fail with a data race when we have unsynchronized access to the stdout, e.g. if multiple things are concurrently writing there. Or, as it happened in the recent integration tests, if trying to read it after an integration test was supposedly done resulted in a data race because some goroutine wasn't done and was still trying to write. 

These are super valuable revelations we are **masking**, not actually fixing, by using locking here.

Reply:
As mentioned on Slack, this synchronizes writing with reading, which is only useful for tests. I.e. we never need to inspect the contents of stdout in the real world, so this is unique for tests.

I don't think we're masking anything with this. We we're manually locking stdout before so we can read from it, and this simply pushes it down the stack so it happens transparently for the test.

Reply:
The fact is that whenever a test wants to inspect the contents of `stdout` or `stderr` when it knows they might be written to (i.e. in the middle of the test), it can manually lock them. But when it wants to inspect them after the test is supposed to be finished, as is the case for the majority of assertions in the integration tests, it's useful to access them directly, without locking. 

That way, if there is a data race, we will be able to see it, and it will most likely be a great signal that we have an actual bug. That is, we want the test to fail! For example, if some forgotten goroutine that is still active and logs stuff after the test is supposed to be finished, as was the case for the `FIXME` comment you deleted. Discovering that data race prompted me to open https://github.com/grafana/k6/issues/2790. And this bug was not fixed in this PR, so you shouldn't have deleted the `FIXME`, it was just masked by this `safeBuffer`. The bug was actually fixed by https://github.com/grafana/k6/pull/2800, which [deleted the `stdout` locking and `FIXME` comment](https://github.com/grafana/k6/pull/2800/files#diff-76b37580ac2865dc962ced55720a7e30bcb354318fee75d8610dd13a92ba2149L1093).

Reply:
OK, understood. I saw the fix in #2800, but didn't think it was a big deal to handle it this way.

How do you want to proceed? Another reason I added `safeBuffer` is because the stdout mutex is no longer exposed to `GlobalState` after the changes here, so tests wouldn't have access to it. I also prefer not having to deal with this locking in the many tests here that inspect stdout while the test is running.

I won't have time to look into this today, but feel free to create a PR if you have a good idea for handling this.

Reply:
I've started working on a PR, but it's likely going to be ready after the holidays. 

> many tests here that inspect stdout while the test is running.

Are there really that many tests? There are 5 `outMutex.Lock()` calls that are removed here, one of which is the `FIXME` under my other comment, another is fixed by another PR, one seems unnecessary (in `TestPrometheusRemoteWriteOutput()`) and the last 2 are combined in a helper function by https://github.com/grafana/k6/pull/2800/commits/d5124d12bbefc028f8e8ff5481abfc492eb05dbc

So, by my reading, after we fix the actual bugs, there will be 1 or 2 necessary lockings the stdout mutex in the integration tests, used only in the few places where we actually care about the stdout contents mid-test.